compiler/rustc_type_ir/src/fold.rs

```diff
@@ -237,3 +237,29 @@ where
         Ok(self.fold_predicate(p))
     }
 }
+
+pub trait SpecTypeFoldable<I: Interner>: Sized {
+    type Item;
+    fn spec_try_fold_with<F: FallibleTypeFolder<I>>(
+        self,
+        folder: &mut F,
+    ) -> Result<Self::Item, F::Error>;
+}
+
+impl<I: Interner, T: SkipTraversalAutoImplOnly> SpecTypeFoldable<I> for &mut Option<T> {
+    type Item = T;
+
+    #[inline(always)]
+    fn spec_try_fold_with<F: FallibleTypeFolder<I>>(self, _: &mut F) -> Result<T, F::Error> {
+        Ok(self.take().unwrap())
+    }
+}
+
+impl<I: Interner, T: TypeFoldable<I>> SpecTypeFoldable<I> for &mut &mut Option<T> {
+    type Item = T;
+
+    #[inline(always)]
+    fn spec_try_fold_with<F: FallibleTypeFolder<I>>(self, folder: &mut F) -> Result<T, F::Error> {
+        self.take().unwrap().try_fold_with(folder)
```

Comment:
I had used `&mut &mut ManuallyDrop<T>` instead, with `unsafe fn spec_try_fold_with` that then performed `ManuallyDrop::take`, but figured because these methods are `#[inline(always)]` directly where the `Option<T>` is constructed as `Some`, the whole wrapping/unwrapping wouldl get optimised away without worrying about the possibility of someone reusing the `ManuallyDrop` after the call.  I was just about to sit down and investigate whether this is indeed the case, or the root of the perf regression however, so your comment/suggestion is well-timed!

I did originally try to use something like your proposed solution, but for some reason had some ownership issues... can't see why now.  Will give it a go.

Reply:
Ah, yeah, I think it's because the deref specialisation doesn't work if `T` is a trait parameter rather than a parameter of the implementing type.

Reply:
Basically, it finds a matching method on `&&()` but fails because type bounds aren't met; whereas in my case there's no such method on `&mut &mut Option<T>`.

Reply:
ah no, I think the issue is that method selection does not look at function arguments:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=04ed2488ae90afacfe5589a3ccc8b46a

you don't need to explicitly write the type of `T` by using the following https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6b928b5b24addc9315fe2483bc1d3c09

so that should be usable

Reply:
Actually it doesn't work if the method takes `T` other than through `self`: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=39cd25a15799d65c90b074e0cd7e6b0d).  Can't say I entirely understand Rust's method resolution here.

Reply:
Okay, but [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=33803622c316373b302d95ebed2afed0) works... feels like a bug somewhere... but I should be able to build a solution on this.

Reply:
Argh, GitHub only just showed your comment!  Sorry, looks like we arrived at the same place :)

Reply:
Okay, that's all done.

Let's see if it made any difference to perf?